### PR TITLE
Iox #714 call reset in subscriber destructor

### LINKED ISSUE #714: Runtime Error in SubscriberImpl
## Required information

**Observed result or behaviour:**
When a to a listener/waitset attached Subscriber (`SubscriberImpl<T>`) goes out of scope it detaches itself in the BaseSubscriber destructor.
1. SubscriberImpl dtor is called and since it inherited from BaseSubscriber it is not a SubscriberImpl anymore after this step
2. BaseSubscriber dtor is called which calls reset an all attached triggers. Since the SubscriberImpl was attached to the Listener/Waitset the trigger reset methods are working on a SubscriberImpl which is no longer a SubscriberImpl since the dtor was already called -> runtime error.

Do not forget the UntypedSubscriberImpl<T>!

**Expected result or behaviour:**
Trigger reset methods are already called in the dtor of SusbcriberImpl.

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
This PR fixes the runtime error described in #714 by calling `m_trigger.reset()` in the destructors of `SubscriberImpl` and `UntypedSubscriberImpl`.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #714

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/popo/subscriber.inl [hunk 1]
```diff
@@ -43,6 +43,12 @@ SubscriberImpl<T, H, BaseSubscriber_t>::take() noexcept
     return cxx::success<Sample<const T, const H>>(std::move(samplePtr));
 }
 
+template <typename T, typename H, typename BaseSubscriber_t>
+SubscriberImpl<T, H, BaseSubscriber_t>::~SubscriberImpl()
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/subscriber.inl:47]: inline missing

File: iceoryx_posh/include/iceoryx_posh/internal/popo/untyped_subscriber.inl [hunk 2]
```diff
@@ -47,6 +47,12 @@ inline void UntypedSubscriberImpl<BaseSubscriber_t>::release(const void* const u
     port().releaseChunk(chunkHeader);
 }
 
+template <typename BaseSubscriber_t>
+UntypedSubscriberImpl<BaseSubscriber_t>::~UntypedSubscriberImpl()
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/untyped_subscriber.inl:51]: inline missing

File: iceoryx_posh/include/iceoryx_posh/popo/subscriber.hpp [hunk 3]
```diff
@@ -38,7 +38,7 @@ class SubscriberImpl : public BaseSubscriber_t
     SubscriberImpl& operator=(const SubscriberImpl&) = delete;
     SubscriberImpl(SubscriberImpl&& rhs) = delete;
     SubscriberImpl& operator=(SubscriberImpl&& rhs) = delete;
-    virtual ~SubscriberImpl() = default;
+    virtual ~SubscriberImpl();
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/popo/subscriber.hpp:41]: Please add noexcept

File: iceoryx_posh/include/iceoryx_posh/popo/untyped_subscriber.hpp [hunk 4]
```diff
@@ -45,7 +45,7 @@ class UntypedSubscriberImpl : public BaseSubscriber_t
     UntypedSubscriberImpl& operator=(const UntypedSubscriberImpl&) = delete;
     UntypedSubscriberImpl(UntypedSubscriberImpl&& rhs) = delete;
     UntypedSubscriberImpl& operator=(UntypedSubscriberImpl&& rhs) = delete;
-    virtual ~UntypedSubscriberImpl() = default;
+    virtual ~UntypedSubscriberImpl();
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/popo/untyped_subscriber.hpp:48]: Please add noexcept

File: iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+TEST_F(PubSubListener_IntegrationTest, SubscriberGoesOutOfScopeAndDeatchingWorks)
+{
+    m_listener
+        ->attachEvent(*m_subscriber,
+                      iox::popo::SubscriberEvent::DATA_RECEIVED,
+                      iox::popo::createEventCallback(onSampleReceivedCallback))
+        .or_else([](auto) { ASSERT_TRUE(false); });
+
+    m_subscriber.reset();
```
Comment (elfenpiff) [inline iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp:71]: maybe we should comment in both tests that we are testing the undefined behavior sanitizer which will go off when the subscriber goes out of scope and the trigger tries to access the child object after the child dtor is finished and it does no longer exist

**Review by elfenpiff — CHANGES_REQUESTED** [review]

**Review by elfenpiff — DISMISSED** [review]

**Review by elBoberido — DISMISSED** [review]

File: iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    ServiceDescription m_serviceDescr{"Radar", "FrontLeft", "Counter"};
+    std::unique_ptr<Listener> m_listener;
+    std::unique_ptr<Subscriber<int>> m_subscriber;
+    std::unique_ptr<UntypedSubscriber> m_untypedSubscriber;
+};
+} // namespace
+
+/// @note Here we test that the trigger reset methods are called correctly in the d'tor of SubscriberImpl. They must not
+/// be called in the BaseSubscriber d'tor since the SubscriberImpl was attached to the Listener. When it goes out of
+/// scope, the trigger tries to access it but SubscriberImpl does not longer exist. This is catched by the
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp:65]: caught

File: iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+/// UndefinedBehaviorSanitizer.
+TEST_F(PubSubListener_IntegrationTest, SubscriberGoesOutOfScopeAndDeatchingWorks)
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/integrationtests/test_popo_pub_sub_listener.cpp:67]: Detaching

**Review by MatthiasKillat — DISMISSED** [review]

**Review by mossmaurice — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

**Pull request merged**